### DISCUSSION TITLE
HMR Not working using anon function pattern (or something else?)

### DISCUSSION THREAD

**Discussion by willnode** [gatsbyjs/gatsby#32456] [Help] [answered]
### Preliminary Checks

- [X] This issue is not a duplicate. Before opening a new issue, please search existing issues: https://github.com/gatsbyjs/gatsby/issues
- [X] This issue is not a question, feature request, RFC, or anything other than a bug report directly related to Gatsby. Please post those things in GitHub Discussions: https://github.com/gatsbyjs/gatsby/discussions

### Description

Hi, I have Gatsby with Material-UI in my project but somehow HMR doesn't work. I'm using next.

![Animation](https://user-images.githubusercontent.com/20214420/126430240-526f3c05-d7af-409b-a0f1-f55793017820.gif)

The test repo is here: https://github.com/willnode/test-gatsby-hmr-bug

The problem is in this line:

https://github.com/willnode/test-gatsby-hmr-bug/blob/74a5739522ad5994b6ad3dbb46356a401422c3d2/src/pages/index.js#L15

If `withTheme` is discarded, HMR working fine. But my project uses it because I want to inject material-ui theme into every page. 

https://github.com/willnode/test-gatsby-hmr-bug/blob/74a5739522ad5994b6ad3dbb46356a401422c3d2/src/components/theme.js#L17-L23

I don't know a better pattern to get my theme injected but because this one creates a bug I will post this anyway.

### Reproduction Link

https://github.com/willnode/test-gatsby-hmr-bug

### Steps to Reproduce

Try HMR

### Expected Result

HMR works

### Actual Result

HMR didn't work

### Environment

```shell
System:
    OS: Windows 10 10.0.19042
    CPU: (6) x64 AMD Ryzen 5 4500U with Radeon Graphics         
  Binaries:
    Node: 14.17.3 - C:\Program Files\nodejs\node.EXE
    Yarn: 1.22.10 - ~\AppData\Roaming\npm\yarn.CMD
    npm: 6.14.13 - C:\Program Files\nodejs\npm.CMD
  Languages:
    Python: 3.9.5 - /c/Users/willnode/AppData/Local/Programs/Python/Python39/python
  Browsers:
    Chrome: 91.0.4472.124
    Edge: Spartan (44.19041.1023.0), Chromium (91.0.864.70)
  npmPackages:
    gatsby: ^3.11.0-next.1 => 3.11.0-next.1 
  npmGlobalPackages:
    gatsby-cli: 3.9.0
```

**Comment by LekoArts** [comment, marked as answer]
Hi!

This is a limitation of React Fast Refresh: https://github.com/facebook/react/issues/17142 and thus working as expected at the moment. To have your theme wrap every page you can use the `wrapRootElement` API in your `gatsby-browser.js` and `gatsby-ssr.js` file: https://www.gatsbyjs.com/docs/reference/config-files/gatsby-browser/#wrapRootElement

The plugin is also doing this: https://github.com/hupe1980/gatsby-plugin-material-ui

**Reply by willnode** [reply]
Oh, thank you! My mistake I'm not digging the docs deeper.
